modelmapper/client.py

```diff
@@ -0,0 +1,159 @@
+                raise ClientSSHException('SFTP CLient {}'.format(e)) from None
+        finally:
+            ssh.close()
+
+    @classmethod
+    def extract(cls, remote_dirpath, *args, **kwargs):
+        """Exposed function.
+
+        Args:
+            cls (type): Class constructor
+            remote_dirpath (str): Base directory where files are pulled.
+
+        Returns:
+            type: Description of returned object.
+
+        Raises:            ExceptionName: Why the exception is raised.
+
+        """
+        sftp_client = cls(*args, **kwargs)
+        return sftp_client.extract(remote_dirpath)
+
+    def _extract(self, remote_dirpath, localpath, historical=True):
+        get_fn = self.get if isinstance(localpath, str) else self.getfo
+        diff = self.contents.difference(self.seen_keys)
+
+        if len(diff) > 0:
+            new_file = sorted(diff)[0]
+            self.logger.info('Files not already extracted: {}'.format(diff))
+
+        return get_fn('{}/{}'.format(remote_dirpath, new_file)), new_file
+
+    @cached_property
+    def contents(self, remotepath):
+        """Lists contents of SFTP at given path.
+
+        Args:
+            remotepath (str): Location on FTP server.
+
+        Returns:
+            list[str]: SFTP directory contents.
+        """
+        with self.get_sftp() as sftp:
+            return sftp.listdir(remotepath)
+
+    def getfo(self, remotepath, file_like_obj, callback=None):
+        """Wrapper around Paramiko.SFTPClient.getfo().
+
+        Args:
+            remotepath (str): Full path on FTP to get
+            file_like_obj (One of (BytesIO, StringIO)): Where to write raw bytes.
+            callback (fn): Custom function
+
+        Returns:
+            file_like_obj (One of (BytesIO, StringIO)): Where to write raw bytes.
+        """
+        callback = callback or self.default_callback
+
+        with self.get_sftp() as sftp:
+
+            self.logger.info('Extracting {}'.format(remotepath))
+            bytes_read = sftp.get(remotepath, file_like_obj, callback=callback)
+
+            if bytes_read < 1:
+                self.logger.info('SFTP did not transfer any data')
+
+            return file_like_obj
+
+    def get(self, remotepath, localpath, callback=None):
+        """Wrapper around Paramiko.SFTPClient.get()."""
+        callback = callback or self.default_callback
+
+        with self.get_sftp() as sftp:
+            bytes_read = sftp.get(remotepath, localpath, callback=callback)
+            if bytes_read < 1:
+                self.logger.info('SFTP did not transfer any data')
+
+    def default_callback(self):
+        """Default callback for all our wrapped Paramiko calls.
+        Possibly dangerous because it relies on the underlying logger
+        to have a info functiond defined.
+        """
+        def cb(logger, seen, total):
+            logger.info('Read {} bytes of total {} bytes.'.format(seen, total))
+        return partial(cb, logger=self.logger)
```

Comment:
I'm not sure why you need the partial here. Why not just:

```
def cb(seen, total):
      self.logger.info... ?
```

I think closure should take care of passing safe.